cohortextractor/dsl.py

```diff
@@ -0,0 +1,178 @@
+"""Skeleton classes and types for the DSL.
+
+This module contains classes and types for building a cohort.
+
+An example of the DSL in use is in cohortextractor/dsl_example.py.
+
+Data lives in database tables, which are represented by PatientTable (a table with one
+row per patient) and EventTable (a table with multiple rows per patient).
+
+Through filtering, sorting, aggregating, and selecting columns, we transform instances
+of PatientTable/EventTable into instances of PatientSeries.
+
+A PatientSeries represents a mapping between patients and values, and can be assigned to
+a Cohort. A PatientSeries can also be combined with another PatientSeries or a scalar
+value to produce a new PatientSeries.
+
+All classes except Cohort are intended to be immutable.
+
+Methods are designed so that users can only perform actions that are semantically
+meaningful.  This means that the order of operations is restricted.  In a terrible ASCII
+railway diagram:
+
+             (  filter                )*  select_column
+PatientTable ( --------> PatientFrame )  ---------------> PatientSeries
+                            ^                                   ^
+                            |                                   |
+                            | first, last                       |
+                            |                                   |
+                            |                                   |
+                    SortedEventFrame                            |
+                            ^                                   |
+                            |                                   |
+                            | sort_by                           |
+                            |                                   |
+                            |                                   |
+           (  filter              )*        count, exists       |
+EventTable ( --------> EventFrame )  ---------------------------+
+
+To support providing helpful error messages, we can implement __getattr__ on each class.
+This will intercept any lookup of a missing attribute, so that if eg a user tries to
+select a column from a SortedEventFrame, we can tell them they need to aggregate the
+SortedEventFrame first.
+"""
+
+from __future__ import annotations
+
+from typing import NoReturn, Union
+
+
+class Cohort:
+    """Represents the cohort of patients in a study."""
+
+    def set_population(self, population: PatientSeries) -> None:
+        """Set the population variable for this cohort."""
+
+    def add_variable(self, name: str, variable: PatientSeries) -> None:
+        """Add a variable to this cohort by name."""
+
+    def __setattr__(self, name: str, variable: PatientSeries) -> None:
+        return self.add_variable(name, variable)
+
+
+class EventFrame:
```

Comment:
Do you anticipate having to write parallel table/frame/sorted-frame classes for each table? Or can we work some type-parameterization magic? Or something else?

Reply:
I'm not sure this would necessary, but I need to do some implementing.